### PR TITLE
Filter engines in application directory so they are not loaded

### PR DESCRIPTION
**Pull request by wildmaples**
### Motivation

Solves: https://github.com/Shopify/tapioca/issues/993

Also fixes an internal Shopify/shopify issue 

### Implementation

As Rafael suggested in https://github.com/Shopify/tapioca/issues/993#issuecomment-1156876059, add a filter to reject engines that are part of the application directory 

### Tests

Change is part of the application loading - can this be tested?

### CODE REVIEW

File: lib/tapioca/runtime/loader.rb [hunk 1]
```diff
@@ -53,9 +53,13 @@ def rails_engines
...
 
         safe_require("active_support/core_ext/class/subclasses")
+        root = File.expand_path(".")
 
         # We can use `Class#descendants` here, since we know Rails is loaded
-        Object.const_get("Rails::Engine").descendants.reject(&:abstract_railtie?)
+        Object.const_get("Rails::Engine")
+          .descendants
+          .reject(&:abstract_railtie?)
+          .reject { |engine| File.fnmatch?("#{root}/**", engine.config.root) }
```
Comment (paracycle) [inline lib/tapioca/runtime/loader.rb:62]: I am afraid this is not enough to check if a gem lives under the application or not. The complications are for gems that are installed under the `vendor` folder inside the application folder, for example.

`Tapioca::Gemfile::GemSpec` class has the method `gem_in_app_dir?`, for example, that checks if any gem files are inside the directory that contains the `Gemfile`, which would also be appropriate here. That method explicitly excludes cases where a given path is under the `Bundler.bundle_path` or `Bundler.app_cache` folders.

File: lib/tapioca/runtime/loader.rb [hunk 2]
```diff
@@ -53,9 +53,13 @@ def rails_engines
         return [] unless Object.const_defined?("Rails::Engine")
 
         safe_require("active_support/core_ext/class/subclasses")
+        root = File.expand_path(".")
```
Comment (paracycle) [inline lib/tapioca/runtime/loader.rb:56]: As mentioned above, a better heuristic for the root folder is to either use `Rails.root` or `Bundler.default_gemfile`.

**Review by paracycle — CHANGES_REQUESTED** [review]

**Comment by wildmaples** [comment]
So I duplicated `#gem_in_app_dir?` from Gemspec in https://github.com/Shopify/tapioca/pull/995/commits/2f0d6a61a7e2775e038ec7ef2fbaa8c9c6765003 because I wanted to confirm the approach works.  

This is, however, duplicated code.`Loader` is does not have convenient access to `Gemspec` unfortunately, and I don't think it's a good idea for it to have access to the instances of `Gemspec`. Both classes have different jobs and are part of different stages in Tapioca's execution. With that, another option is to extract the logic into a utility module that can be used by both `Loader` and `Gemspec`. What do you think?

**Comment by rafaelfranca** [comment]
> With that, another option is to extract the logic into a utility module that can be used by both `Loader` and `Gemspec`

I think that is a valid way to share code. If you can use composition for this it would be better in my opinion but also totally fine to use inheritance.

**Comment by Morriar** [comment]
I think we can extract those methods under a new `helpers/gem_helper` module and include it in both places.

@rafaelfranca for composition are you suggesting making it a class that is instantiated in both places? It doesn't seem like those methods require any state?

**Review by Morriar — APPROVED** [review]
> `tapioca gem --all` runs successfully.

It's also good to check if `srb tc` finished without any unexpected errors. This may be the sign we're missing some constants that should be generated.

**Comment by wildmaples** [comment]
Thanks for the review! I ran `srb tc` and there are errors so I'm gonna investigate to see if they are real changes before merging this.

**Comment by KaanOzkan** [comment]
@DavidPD Since we don't support Ruby 2.6 (EOL) I think this would set a bad precedent for the repo. However, I'm confident that we won't backport anything else to `0.7.x` so if you were to point to your personal fork you wouldn't be at a disadvantage compared to sticking to tapioca `0.7.x`.

So I suggest you fork the project, backport the fix and point tapioca to your fork. This might even be the better long term solution if you're sticking to Ruby 2.6 for a while since you won't be dependent on us for other backports. In the future when you can upgrade Ruby you can point back to RubyGems releases.

I'm very sorry about the inconvenience, hopefully you can utilize latest Tapioca in no time since there have been a ton of improvements 🙂

**Comment by DavidPD** [comment]
@KaanOzkan Thanks for the update. Yeah that's pretty much what we're doing short-term, and hopefully not _too_ far down the road we can get our project somewhat up-to-date.

**Pull request merged**